Updated SearchDelegate to follow custom InputDecorationTheme

gildaswise opened the pull request:
## Description

Currently, if you start using `InputDecorationTheme` in your app, `SearchDelegate` will follow it and work as intended, but if you're using something like a `OutlineInputBorder`, it fails to display it properly, and even overwriting `appBarTheme` with another `InputDecorationTheme` doesn't help, because the current version of `SearchDelegate` does not apply it as a `Theme` widget.

This PR makes it so `AppBar` is inside a `Theme` widget that applies the `appBarTheme` property to the whole tree. It also moves the `InputDecorationTheme` changes to that property, so the only thing in `decoration:` is `hintText`, and every other theming property should be added to `appBarTheme` or `searchFieldDecoration` instead of directly there.

## Related Issues

- https://github.com/flutter/flutter/issues/52483
- https://github.com/flutter/flutter/issues/54673
- https://github.com/flutter/flutter/issues/45498
- https://github.com/flutter/flutter/issues/48857

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.

gildaswise committed:
Updated SearchDelegate to follow InputDecorationTheme

gildaswise committed:
Merge branch 'master' into gildaswise-search-delegate

gildaswise commented:
Just a heads-up, this is different from [#56577](https://github.com/flutter/flutter/pull/56577), that one just adds more properties coming from `theme.inputDecorationTheme` instead of completely using it, as it should be. This one fixes the issue of a changed `InputDecoration` not working at all with the `SearchDelegate`, and makes it so we don't have to keep adding properties manually like that.

hawkinsjb1 commented:
What's the status of this? The current `SearchDelegate` implementation is very messy when it comes to custom themes because it tries to cherry-pick specific theme params as opposed to wrapping in a `ThemeData` like this fix does

gildaswise commented:
@hawkinsjb1 Still waiting on something too, I believe it's still up-to-date with the latest delegate.

darrenaustin reviewed (changes requested):
@gildaswise, so sorry it has taken me a while to get to this. Thanks for your contribution.

It looks like this has been here long enough to get out of sync with the master branch, so this will need to be updated.

The change itself looks reasonable, but we will also need a test added that verifies that this works so that no one will break this in the future.

If you are no longer interested in working on this, just let us know and we can find someone to take it on. Thx.

gildaswise commented:
@darrenaustin Starting to work on this again right now, thanks for looking into it!

gildaswise committed:
Merge pull request #1 from flutter/master

Rebase branch

gildaswise committed:
Added 'Custom inputDecorationTheme value' test

gildaswise committed:
Added 'Custom inputDecorationTheme value' test

gildaswise committed:
Re-modified search.dart, fixing branch

gildaswise committed:
Still trying to rebase

gildaswise committed:
Reworked PR, added searchFieldDecoration and updated tests to use ThemeData instead of raw InputDecoration

gildaswise committed:
Removed leftover print statements

gildaswise commented:
@darrenaustin Just finished, had to change my approach to this. Before, it was meant to be a PR that fixed custom `InputDecorationTheme` usage when that's defined way back in `MaterialApp`'s theme property, but that messed things up when you tried to just change something manually, or wanted to only customize a single screen. 

I changed it to add the `searchFieldDecoration` property that should handle all `InputDecorationTheme` stuff and still manages to apply your default `InputDecorationTheme` if available. 

I had to update two tests that relied on checking `TextField`'s `decoration` property, but since I changed it to only use the Theme version, they started to fail searching for properties that aren't there anymore. So now those tests and the new one check the Widget's ThemeData instead.

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -274,6 +280,13 @@ abstract class SearchDelegate<T> {
+  ///
+  /// If this value is not null, [searchFieldStyle]
+  /// will be ignored so this can be used.
+  final InputDecorationTheme searchFieldDecoration;
```
I would prefer naming this `searchFieldDecorationTheme` to avoid confusion with regular `InputDecoration` and `InputDecorator` classes.

Also, if we just ignore `searchFieldStyle` if this is set, perhaps we should have an assert in the constructor that ensures only one of these properties is set.

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -476,12 +493,10 @@ class _SearchPageState<T> extends State<_SearchPage<T>> {
-      ?? MaterialLocalizations.of(context).searchFieldLabel;
-    final TextStyle searchFieldStyle = widget.delegate.searchFieldStyle
-      ?? theme.inputDecorationTheme.hintStyle;
+    final String searchFieldLabel = widget.delegate.searchFieldLabel ??
```
It is unfortunate that the delegate function was called `appBarTheme`. It really is a full `ThemeData` and with the changes below it will be effecting everything in the search UI, not just the app bar. Not sure of a better solution here though short of renaming it, so this will work for now.

gildaswise replied:
That's basically the whole issue on this, since the `ThemeData` is not being properly applied by default, there's all sorts of inconsistencies on the `SearchDelegate`. But the way this is working, `appBarTheme` should always come from a `Theme.of(context)` either way, so unless someone REALLY applies another entirely new theme there, we shouldn't have issues on this approach.

darrenaustin reviewed (changes requested):
Sorry it took a while to get to this. Thanks so much for the contribution. There are several PRs that are attempting to solve various parts of this issue, but yours seems the most robust. I have a few issues though.

First, there are a lot of unrelated style changes here that make it hard to see what the actual fix is that you are proposing. The Flutter framework uses a different [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) from dart. Can you please revert these style changes to make it easier to review your fix?

I have other code specific comments below. Thanks again.

gildaswise committed:
fix: fixed formatting issues, updated requested changes

gildaswise committed:
fix: more minor formatting corrections

gildaswise committed:
fix: resolved merge conflict

gildaswise committed:
Merge branch 'master' into gildaswise-search-delegate

gildaswise commented:
@darrenaustin Finished working on it now, as I did `git fetch upstream` and solved the merge conflict, there was an issue where the linter was complaining that `routeName` wasn't being set, so I changed that `switch` to fix that. If that's not in the scope, please let me know to remove it.

Sorry on the style issues, I really didn't pay attention that I shouldn't use `dartfmt` while working on the Flutter repo and I had `formatOnSave` toggled on VSCode. Let me know if you need anything else!

gildaswise committed:
fix: even more formatting fixes

gildaswise committed:
fix: updated copyright header

gildaswise committed:
Merge branch 'master' into gildaswise-search-delegate

gildaswise committed:
fix: fixed nullability markers

darrenaustin reviewed (changes requested):
@gildaswise, thanks again for continuing to work on this PR. Apologies it has taken me a while to review your latest updates. It still looks like a fair amount of changes here that aren't related to just fixing the problem. I have noted them below.

There are also a couple of suggestions I have for the tests. 

Otherwise this looks good. We have an internal user that is keen to get this fix, so I would like to expedite getting this merged. If you don't have the time or desire to continue with this, I can take it from here and build a new PR with what we need (giving you credit of course). Just let me know. Thanks again!

gildaswise commented on packages/flutter/test/material/search_test.dart:
```diff
@@ -129,10 +129,10 @@ void main() {
-    final Color hintColor = textField.decoration!.hintStyle!.color!;
-    expect(hintColor, delegate.hintTextColor);
+    
+    final ThemeData? textFieldTheme = Theme.of(tester.element(find.byType(TextField)));
```
Ooh, I didn't know that getting the text directly would also get the style like that. I changed to match the `InputDecorationTheme` check because `hintStyle` was returning null (since we're applying a theme now instead of a direct `InputDecoration`). Will update the test, thanks.

gildaswise commented on packages/flutter/test/material/search_test.dart:
```diff
@@ -540,19 +540,17 @@ void main() {
 
-    final TextField textField = tester.widget<TextField>(find.byType(TextField));
-    final TextStyle hintStyle = textField.decoration!.hintStyle!;
+    final ThemeData? textFieldTheme = Theme.of(tester.element(find.byType(TextField)));
```
Had to tweak something for this one, comparing the `TextStyle` directly returned this since it's already rendered:

```
The following TestFailure object was thrown running a test:
  Expected: TextStyle:<TextStyle(inherit: true, color: MaterialColor(primary value:
Color(0xfff44336)), size: 3.0)>
  Actual: TextStyle:<TextStyle(debugLabel: (((englishLike title 2014).merge(blackMountainView
headline6)).copyWith).merge(unknown), inherit: false, color: MaterialColor(primary value:
Color(0xfff44336)), family: Roboto, size: 3.0, weight: 500, baseline: alphabetic, decoration:
TextDecoration.none)>
```

So I changed the test to only check the altered values instead:

```
final Text hintText = tester.widget(find.text(searchHintText));
expect(hintText.style?.color, delegate.searchFieldStyle?.color);
expect(hintText.style?.fontSize, delegate.searchFieldStyle?.fontSize);
```

gildaswise committed:
Removed trailing spaces and unrelated changes, updated tests to check Text directly

gildaswise committed:
Removed more trailing spaces

gildaswise committed:
Removed more unrelated changes due to auto format

gildaswise committed:
Removed more unrelated changes due to auto format

gildaswise commented:
> @gildaswise, thanks again for continuing to work on this PR. Apologies it has taken me a while to review your latest updates. It still looks like a fair amount of changes here that aren't related to just fixing the problem. I have noted them below.
> 
> There are also a couple of suggestions I have for the tests.
> 
> Otherwise this looks good. We have an internal user that is keen to get this fix, so I would like to expedite getting this merged. If you don't have the time or desire to continue with this, I can take it from here and build a new PR with what we need (giving you credit of course). Just let me know. Thanks again!

Finally removed all unrelated changes, so sorry about that. I checked the file diff and everything should be on point now. Just let me know about the last test comment I made and it should be fine now. Thanks for reviewing and I understand the internal user, haha.

gildaswise committed:
Added changes from #67679

gildaswise committed:
Merged latest from upstream

gildaswise commented:
I just fixed the merge conflict, but the two new tests from #67679 are not working. I tried just removing my changes to see if would go fine, but it didn't. It says that `AppBar.backgroundColor` is null, so it fails on both lines 728 and 753.

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: Color:<Color(0xff212121)>
  Actual: <null>
```

I also just added a quick `print('is appBar.backgroundColor null: ${appBar.backgroundColor}` check before the expect call and: `is appBar.backgroundColor null: null` was the return. Not really sure what's going on here and would appreciate help, @darrenaustin or @justinmc.

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -92,9 +92,8 @@ Future<T?> showSearch<T>({
-
   /// Constructor to be called by subclasses which may specify [searchFieldLabel], [keyboardType] and/or
-  /// [textInputAction].
+  /// [textInputAction]. You can only use either [searchFieldStyle] or [searchFieldDecorationTheme] at a time.
```
Perhaps better wording:

```dart
  /// Constructor to be called by subclasses which may specify
  /// [searchFieldLabel], either [searchFieldStyle] or [searchFieldDecorationTheme],
  /// [keyboardType] and/or [textInputAction]. Only one of [searchFieldLabel]
  /// and [searchFieldDecorationTheme] may be non-null.
```

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -501,51 +514,47 @@ class _SearchPageState<T> extends State<_SearchPage<T>> {
       case TargetPlatform.linux:
       case TargetPlatform.windows:
         routeName = searchFieldLabel;
+        break;
```
All of these changes to the switch statement are unrelated to the decoration changes, so we shouldn't modify it here.

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -501,51 +514,47 @@ class _SearchPageState<T> extends State<_SearchPage<T>> {
         break;
     }
-    final String routeName;
+    String? routeName;
```
I am surprised that this was not already caught when we were updating to null safety, but a preferred way to handle this kind of case would be to have it be declared:

```dart
late finale routeName;
```

To prefer a non-nullable value.

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -275,9 +280,19 @@ abstract class SearchDelegate<T> {
   ///
   /// If this value is set to null, the value of the ambient [Theme]'s
-  /// [InputDecorationTheme.hintStyle] will be used instead.
+  /// [InputDecorationTheme.hintStyle] will be used instead
```
Need to put the '.' back at the end of this line.

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -501,51 +514,47 @@ class _SearchPageState<T> extends State<_SearchPage<T>> {
       label: routeName,
-      child: Scaffold(
-        appBar: AppBar(
-          backgroundColor: theme.primaryColor,
-          iconTheme: theme.primaryIconTheme,
-          textTheme: theme.primaryTextTheme,
-          brightness: theme.primaryColorBrightness,
```
These shouldn't be removed, as they are needed to set the appropriate properties on the AppBar. We could get them from the `AppBarTheme` of the theme, but that is not how the default theme is setup. We are looking at cleaning up the `AppBarTheme` soon, but for now please just put these back. This will fix the new tests that are failing.

darrenaustin reviewed (changes requested):
Making good progress. Thanks. I still have a few requests below. The tests are breaking due to the removal of some settings on the `AppBar`. I have pointed them out below as well.

gildaswise committed:
Comment changes, back with AppBar properties

darrenaustin commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -519,33 +535,32 @@ class _SearchPageState<T> extends State<_SearchPage<T>> {
+        data: theme,
+        child: Scaffold(
+          appBar: AppBar(
+            backgroundColor: theme.primaryColor,
+            iconTheme: theme.primaryIconTheme,
+            textTheme: theme.primaryTextTheme,
+            brightness: theme.primaryColorBrightness,
```
Ok, really sorry about this, but I was wrong here. We do need these removed, or apps won't be able to override them with `AppBarTheme` settings (either globally or in the returned theme from the delegates method). This will break the sample code in #45498.

So please remove this again. We can fix the tests below.

darrenaustin reviewed (changes requested):
Looks good, and the tests are passing. However, I steered you the wrong way on putting those AppBar parameters back. We need to remove them, as you had it right the first time. Really sorry about that.

For the new tests that will break with them missing, we can adjust them to look at the underlying `Material` widget instead of the `AppBar` itself. I can't seem to put a comment on the code itself, but in the test code for 'text in search bar contrasts background (light mode)', you can change the appBar color check to:

```dart
    final Material appBarBackground = tester.widget<Material>(find.descendant(
      of: find.byType(AppBar),
      matching: find.byType(Material),
    ));
    expect(appBarBackground.color, Colors.white);
```

And the equivalent for the dark mode test.

Again, thanks for your patience and work here. I think it is really close.

gildaswise committed:
Removed AppBar parameters, updated tests

darrenaustin reviewed (approved):
![](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

Thanks for all the work on this @gildaswise!

gildaswise commented:
Thanks for reviewing, @darrenaustin! Hope everyone benefits from this asap, too.

gildaswise commented:
@darrenaustin Let me know what exactly I need to change here, unfortunately the "Details" button on Google testing doesn't really help.

darrenaustin commented:
> @darrenaustin Let me know what exactly I need to change here, unfortunately the "Details" button on Google testing doesn't really help.

Yeah, it tripped an internal test. Looking at the results I don't think it is related to your change. I am trying to figure out what is going on and will get back to you and hopefully we can land this soon. Thx.

vasilich6107 commented:
Congrats @gildaswise and @darrenaustin  for not giving up and finishing this PR!

juliavi commented:
Still not resolved in the latest version

intraector commented:
Same issue

justinmc commented:
@juliavi @intraector What Flutter version specifically are you seeing this problem on?  Can you open a new issue with repro instructions?
